# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmValType.cpp
# Commit: 72bf9c871ded
# Full Hash: 72bf9c871ded006f56f09b24d0723365516445bb
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2023-08-05 09:19:01
# Regressor Bug: 1692065
# File Overlap Count: 5
# Description:
#   Bug 1692065 - wasm: Add tagging for 31-bit integers and i31ref. r=yury
#   
#   This commit adds a tag 0x1 for 31-bit integers in AnyRef. Any JS value
#   that is losslessly representable (without coercion) in this range is
#   stored without boxing. This is chosen to line up exactly with how the
# ==============================================================================

diff -r 28feda14fdec -r 72bf9c871ded js/src/wasm/WasmValType.cpp
--- a/js/src/wasm/WasmValType.cpp	Fri Aug 04 14:06:50 2023 +0000
+++ b/js/src/wasm/WasmValType.cpp	Fri Aug 04 14:06:50 2023 +0000
@@ -40,6 +40,7 @@
   switch (kind()) {
     case RefType::Any:
     case RefType::Eq:
+    case RefType::I31:
     case RefType::Array:
     case RefType::Struct:
     case RefType::None:
@@ -101,6 +102,10 @@
       *out = RefType::eq();
       return true;
     }
+    if (StringEqualsLiteral(typeLinearStr, "i31ref")) {
+      *out = RefType::i31();
+      return true;
+    }
     if (StringEqualsLiteral(typeLinearStr, "structref")) {
       *out = RefType::struct_();
       return true;
@@ -172,6 +177,8 @@
     *out = RefType::any();
   } else if (GcAvailable(cx) && StringEqualsLiteral(typeLinearStr, "eq")) {
     *out = RefType::eq();
+  } else if (GcAvailable(cx) && StringEqualsLiteral(typeLinearStr, "i31")) {
+    *out = RefType::i31();
   } else if (GcAvailable(cx) && StringEqualsLiteral(typeLinearStr, "struct")) {
     *out = RefType::struct_();
   } else if (GcAvailable(cx) && StringEqualsLiteral(typeLinearStr, "array")) {
@@ -304,6 +311,9 @@
       case RefType::Eq:
         literal = "eqref";
         break;
+      case RefType::I31:
+        literal = "i31ref";
+        break;
       case RefType::Struct:
         literal = "structref";
         break;
@@ -341,6 +351,9 @@
     case RefType::Eq:
       heapType = "eq";
       break;
+    case RefType::I31:
+      heapType = "eq";
+      break;
     case RefType::Struct:
       heapType = "struct";
       break;